python/ray/dag/dag_node_operation.py

```diff
@@ -268,36 +356,58 @@ def _build_dag_node_operation_graph(
             }
 
     # Import `ray.dag` here to avoid circular import.
-    from ray.dag import ClassMethodNode, MultiOutputNode
+    from ray.dag import ClassMethodNode, CollectiveOutputNode, MultiOutputNode
 
     # Add an edge from WRITE of the writer task to READ of the reader task.
     for task_idx, task in idx_to_task.items():
-        if (
+        if not (
             isinstance(task.dag_node, ClassMethodNode)
-            and task.dag_node.is_class_method_output
+            or isinstance(task.dag_node, CollectiveOutputNode)
         ):
-            # TODO(wxdeng): Handle the case where the task is a class method output.
-            continue
-        if not isinstance(task.dag_node, ClassMethodNode):
             # The graph is used to generate an execution schedule for each actor.
             # The edge from the InputNode has no impact on the final execution
             # schedule.
             continue
+        if (
+            isinstance(task.dag_node, ClassMethodNode)
+            and task.dag_node.is_class_method_output
+        ):
+            # TODO(wxdeng): Handle the case where the task is a class method output.
+            continue
         for downstream_task_idx in task.downstream_task_idxs:
             downstream_dag_node = idx_to_task[downstream_task_idx].dag_node
+            if isinstance(downstream_dag_node, MultiOutputNode):
+                continue
             if (
                 isinstance(downstream_dag_node, ClassMethodNode)
                 and downstream_dag_node.is_class_method_output
             ):
                 # TODO(wxdeng): Handle the case where the downstream task is
                 # a class method output.
                 continue
-            if isinstance(downstream_dag_node, MultiOutputNode):
-                continue
             _add_edge(
                 graph[task_idx][_DAGNodeOperationType.WRITE],
                 graph[downstream_task_idx][_DAGNodeOperationType.READ],
             )
+
+    from ray.dag.collective_node import _CollectiveGroup
+
+    # Set collective group nodes for all the NCCL compute nodes.
+    collective_group_to_task_idxs: Dict[_CollectiveGroup, Set[int]] = defaultdict(set)
```

Comment:
It is more robust to set `collective_group` during the _DAGNodeOperation constructor instead of afterwards.

Reply:
Ack. Done.